sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/JobFailure.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.loadtests;
+
+import static java.lang.String.format;
+import static java.util.Optional.empty;
+import static java.util.Optional.of;
+
+import java.io.IOException;
+import java.util.Optional;
+import org.apache.beam.sdk.PipelineResult;
+import org.joda.time.Duration;
+
+/** Class for detecting failures after {@link PipelineResult#waitUntilFinish(Duration)} unblocks. */
+class JobFailure {
+
+  private String cause;
+
+  private boolean requiresCancelling;
+
+  JobFailure(String cause, boolean requiresCancelling) {
+    this.cause = cause;
+    this.requiresCancelling = requiresCancelling;
+  }
+
+  static Optional<JobFailure> assertFailure(PipelineResult pipelineResult,
+      LoadTestResult testResult) {
+    PipelineResult.State state = pipelineResult.getState();
+
+    Optional<JobFailure> stateRelatedFailure = lookForInvalidState(state);
+
+    if (stateRelatedFailure.isPresent()) {
+      return stateRelatedFailure;
```

Comment:
I'd suggest to throw an `AssertionError` based exception instead of return value. Otherwise, it would be better to rename the method `assertFailure` to avoid confusing with other `assert*`-like methods (e.g. from JUnit).

Reply:
ok - I renamed it. I also encapsulated this stuff in one method. However, I think it's better if we stick with RuntimeException here - it think it fits better for pipeline failures.